transforms3d/euler.py

```diff
@@ -395,13 +399,106 @@ def quat2euler(quaternion, axes='sxyz'):
     ak : float
         Third rotation angle (according to `axes`).
 
+    References
+    ----------
+    .. [1] Bernardes E, Viollet S (2022) Quaternion to Euler angles
+           conversion: A direct, general and computationally efficient
+           method. PLoS ONE 17(11): e0276302.
+           https://doi.org/10.1371/journal.pone.0276302
+
     Examples
     --------
     >>> angles = quat2euler([0.99810947, 0.06146124, 0, 0])
     >>> np.allclose(angles, [0.123, 0, 0])
     True
     """
-    return mat2euler(quat2mat(quaternion), axes)
+    if not direct_formula:
+        angles = list(mat2euler(quat2mat(quaternion), axes))
+        if angles[1] < 0 and axes[1] == axes[-1]:
+            angles[0] = (angles[0]) % (2 * np.pi) - np.pi
+            angles[1] = -angles[1]
+            angles[2] = (angles[2]) % (2 * np.pi) - np.pi
+        return tuple(angles)
+
+    # check if valid axes option
+    axes = axes.lower()
+    _AXES2TUPLE[axes]
+
+    # if direct formula, use the algorithm:
+    extrinsic = axes[0] == 's'
+    seq = axes[1:]
+    if extrinsic:
+        angle_first = 0
+        angle_third = 2
+    else:
+        seq = seq[::-1]
+        angle_first = 2
+        angle_third = 0
+
+    i = 'xyz'.index(seq[0]) + 1
+    j = 'xyz'.index(seq[1]) + 1
+    k = 'xyz'.index(seq[2]) + 1
+
+    symmetric = i == k
+
+    if symmetric:
+        k = 6 - i - j  # get third axis
+
+    # Step 0
```

Comment:
Forgive my ignorance, but is this the same as the `parity` check returned from `_AXES2TUPLE`?

Reply:
No, this is to get the third axis for symmetric sequences.
For example, for the sequence `'zyz'`, at first we have `i, j, k == 3, 2, 3`.
The third axis (the axis we don't have) is `1`. Since `i, j, k` can only be `1`, `2` or `3`, we know that `i + j + k_new == 6`.
So: `k_new = 6 - i - j` gets this other axis.